ethdb/olddb/memorymutation.go

```diff
@@ -183,109 +183,155 @@ func (m *miningmutation) Has(table string, key []byte) (bool, error) {
 	return m.memTx.Delete(table, k, v)
 }
 
-func (m *miningmutation) Commit() error {
+func (m *memorymutation) Commit() error {
 	return nil
 }
 
-func (m *miningmutation) Rollback() {
+func (m *memorymutation) Rollback() {
 	m.memTx.Rollback()
 	m.memDb.Close()
 	return
 }
 
-func (m *miningmutation) Close() {
+func (m *memorymutation) Close() {
 	m.Rollback()
 }
 
-func (m *miningmutation) Begin(ctx context.Context, flags ethdb.TxFlags) (ethdb.DbWithPendingMutations, error) {
+func (m *memorymutation) Begin(ctx context.Context, flags ethdb.TxFlags) (ethdb.DbWithPendingMutations, error) {
 	panic("mutation can't start transaction, because doesn't own it")
 }
 
-func (m *miningmutation) panicOnEmptyDB() {
+func (m *memorymutation) panicOnEmptyDB() {
 	if m.db == nil {
 		panic("Not implemented")
 	}
 }
 
-func (m *miningmutation) SetRwKV(kv kv.RwDB) {
+func (m *memorymutation) SetRwKV(kv kv.RwDB) {
 	m.db.(ethdb.HasRwKV).SetRwKV(kv)
 }
 
-func (m *miningmutation) BucketSize(bucket string) (uint64, error) {
+func (m *memorymutation) BucketSize(bucket string) (uint64, error) {
 	return 0, nil
 }
 
-func (m *miningmutation) DropBucket(bucket string) error {
+func (m *memorymutation) DropBucket(bucket string) error {
 	panic("Not implemented")
 }
 
-func (m *miningmutation) ExistsBucket(bucket string) (bool, error) {
+func (m *memorymutation) ExistsBucket(bucket string) (bool, error) {
 	panic("Not implemented")
 }
 
-func (m *miningmutation) ListBuckets() ([]string, error) {
+func (m *memorymutation) ListBuckets() ([]string, error) {
 	panic("Not implemented")
 }
 
-func (m *miningmutation) ClearBucket(bucket string) error {
+func (m *memorymutation) ClearBucket(bucket string) error {
 	m.clearedTables[bucket] = struct{}{}
 	return m.memTx.ClearBucket(bucket)
 }
 
-func (m *miningmutation) isBucketCleared(bucket string) bool {
+func (m *memorymutation) isBucketCleared(bucket string) bool {
 	_, ok := m.clearedTables[bucket]
 	return ok
 }
 
-func (m *miningmutation) CollectMetrics() {
+func (m *memorymutation) CollectMetrics() {
 }
 
-func (m *miningmutation) CreateBucket(bucket string) error {
+func (m *memorymutation) CreateBucket(bucket string) error {
 	panic("Not implemented")
 }
 
+func (m *memorymutation) Flush(tx kv.RwTx) error {
+	// Obtain buckets touched.
+	buckets, err := m.memTx.ListBuckets()
+	if err != nil {
+		return err
+	}
+	// Iterate over each bucket and apply changes accordingly.
+	for _, bucket := range buckets {
+		if _, ok := m.dupsortTables[bucket]; ok && bucket != kv.HashedStorage {
+			cbucket, err := m.memTx.CursorDupSort(bucket)
+			if err != nil {
+				return err
+			}
+			defer cbucket.Close()
+			dbCursor, err := tx.RwCursorDupSort(bucket)
+			if err != nil {
+				return err
+			}
+			defer dbCursor.Close()
+			for k, v, err := cbucket.First(); k != nil; k, v, err = cbucket.Next() {
+				if err != nil {
+					return err
+				}
+				if err := dbCursor.AppendDup(k, v); err != nil {
```

Comment:
1. probably auto-dupsort tables need exclude (not sure)
2. it can work only for "appendable" tables (like changeSets), if db already has some key greater than this one (for example state storage) .AppendDup will return error. 
3. You can see in etl - how it choose between append and put.

Reply:
dupsorted tables are excluded, and we only care about changesets in this case. this is to be used to keep in-memory state, but can be fixed.